File: content/problems/2680-maximum-or/solutions/enumerate-shifted-index-with-suffix-or/Solution.kt
class Solution {
    fun maximumOr(nums:IntArray,k:Int):Long{
        val suffix=LongArray(nums.size+1);
        for(i in nums.lastIndex downTo 0)suffix[i]=suffix[i+1] or nums[i].toLong();
        var prefix=0L;
        var answer=0L;
        for(i in nums.indices){
            answer=maxOf(answer,prefix or (nums[i].toLong() shl k) or suffix[i+1]);
            prefix=prefix or nums[i].toLong()
        };
        return answer
    }
}

File: content/problems/2680-maximum-or/solutions/enumerate-shifted-index-with-suffix-or/solution.cpp
class Solution {
    public:long long maximumOr(vector<int>& nums,int k){
        vector<long long> suffix(nums.size()+1);
        for(int i=nums.size()-1;i>=0;--i)suffix[i]=suffix[i+1]|nums[i];
        long long prefix=0,answer=0;
        for(int i=0;i<(int)nums.size();++i){
            answer=max(answer,prefix|(1LL*nums[i]<<k)|suffix[i+1]);
            prefix|=nums[i];
        }
        return answer;
    }
};
